<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Welcome to fwdpy’s documentation! &#8212; fwdpy 0.0.4-rc3 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0.4-rc3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="fwdpy: forward simulation in Python using fwdpp" href="README.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="README.html" title="fwdpy: forward simulation in Python using fwdpp"
             accesskey="N">next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">fwdpy 0.0.4-rc3 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="welcome-to-fwdpy-s-documentation">
<h1>Welcome to fwdpy&#8217;s documentation!<a class="headerlink" href="#welcome-to-fwdpy-s-documentation" title="Permalink to this headline">¶</a></h1>
<p>Contents:</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="README.html">fwdpy: forward simulation in Python using fwdpp</a><ul>
<li class="toctree-l2"><a class="reference internal" href="README.html#citation">Citation</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#features">Features:</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#google-group">Google Group</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#availability">Availability</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#dependencies">Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#what-python-version">What Python version?</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#testing">Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#note-for-developers">Note for developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="README.html#documentation">Documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/changelog.html">Changelog (rough)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/changelog.html#rc3">0.0.4-rc3</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/changelog.html#through-release-candidate-2">0.0.4 (through release candidate 2)</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/changelog.html#id1">0.0.3</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/todo.html">TODO</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/notes.html">Modules in this package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/notes.html#notes-on-the-documentation">Notes on the documentation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/rusers.html">A note for R users</a></li>
</ul>
</div>
<div class="section" id="theoretical-concepts">
<h2>Theoretical concepts<a class="headerlink" href="#theoretical-concepts" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/background.html">Fitness models</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/background.html#theoretical-details">Theoretical details</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/background.html#the-relationship-between-forward-time-and-backward-time-coalescent-simulations">The relationship between forward-time and backward-time (coalescent) simulations.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/background.html#references">References</a></li>
</ul>
</div>
</div>
<div class="section" id="important-concepts-for-setting-up-simulations">
<h2>Important concepts for setting up simulations<a class="headerlink" href="#important-concepts-for-setting-up-simulations" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/popsizes.html">Example: modeling changes in population size</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/popsizes.html#simple-example">Simple example</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/popsizes.html#a-simple-bottleneck">A simple bottleneck</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/popsizes.html#exponential-growth">Exponential growth</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/popsizes.html#potential-caveat">Potential caveat</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/popsizes.html#rationale">Rationale</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/demography.html">Complex demographic operations</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/regions.html">Example: Modeling regional variation in mutation and recombination</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/regions.html#background">Background</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/regions.html#specific-examples">Specific examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/regions.html#how-to-set-up-a-model">How to set up a model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/summstats.html">Calculating summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/slim.html">Processing SLiM input files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/slim.html#differences-between-slim-and-fwdpy">Differences between SLiM and fwdpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/slim.html#a-word-of-caution">A word of caution</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/slim.html#limitations">Limitations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/threads.html">How many threads to use?</a></li>
</ul>
</div>
</div>
<div class="section" id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="examples/BGS.html">Example: background selection</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples/BGS.html#setting-up-the-simulation">Setting up the simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples/BGS.html#taking-samples-from-simulated-populations">Taking samples from simulated populations</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples/BGS.html#summary-statistics-from-samples">Summary statistics from samples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples/views.html">Example of taking &#8216;views&#8217; from simulated populations</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/windows.html">Example: Sliding windows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples/windows.html#calculating-sliding-windows">Calculating sliding windows</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples/trajectories.html">Tracking mutation frequencies</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples/trajectories.html#run-a-simulation">Run a simulation</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples/trajectories.html#group-mutation-trajectories-by-position-and-effect-size">Group mutation trajectories by position and effect size</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples/copypops.html">Copying populations in memory</a></li>
</ul>
</div>
</div>
<div class="section" id="special-topics">
<h2>Special topics<a class="headerlink" href="#special-topics" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/vareg.html">Estimating cumulative <span class="math">\(V(A)\)</span> in simulations of quantitative traits.</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/vareg.html#technical-notes">Technical Notes</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/vareg.html#references">References</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="plugins">
<h2>Plugins<a class="headerlink" href="#plugins" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/plugins.html">Overview</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#glossary">Glossary</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#file-extensions">File extensions</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#compiling-a-cython-only-plugin">Compiling a Cython-only plugin</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#compiling-a-plugin-that-contains-extra-c-code">Compiling a plugin that contains extra C++ code</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#the-c-types-used-in-fwdpy">The C++ types used in <em>fwdpy</em></a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#additional-material">Additional material</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="pages/plugins.html#implementing-custom-fitness-functions">Implementing custom fitness functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/plugins.html#concepts">Concepts</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="notes-for-people-wanting-to-work-with-the-code">
<h2>Notes for people wanting to work with the code<a class="headerlink" href="#notes-for-people-wanting-to-work-with-the-code" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pages/code.html">Source code organization</a></li>
<li class="toctree-l1"><a class="reference internal" href="pages/extending.html">Extending <em>fwdpy</em></a><ul>
<li class="toctree-l2"><a class="reference internal" href="pages/extending.html#including-the-c-headers">Including the C++ headers</a></li>
<li class="toctree-l2"><a class="reference internal" href="pages/extending.html#api-documentation">API Documentation</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="module-fwdpy.fwdpy">
<span id="functions-and-classes"></span><h2>Functions and classes<a class="headerlink" href="#module-fwdpy.fwdpy" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="fwdpy.fwdpy.ConstantS">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">ConstantS</code><a class="headerlink" href="#fwdpy.fwdpy.ConstantS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></p>
<p>Constant/fixed selection coefficient</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>s: the selection coefficient</li>
<li>h: the dominance term</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation</p>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>s</strong> &#8211; the selection coefficient</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#s = -0.1 and h = 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constantS</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">ConstantS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.ExpS">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">ExpS</code><a class="headerlink" href="#fwdpy.fwdpy.ExpS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></p>
<p>Exponential distribution on selection coefficients</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>mean: the mean selection coefficient</li>
<li>h: the dominance term</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation</p>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>mean</strong> &#8211; the mean selection coefficient</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#s is exp(-0.1) and recessive</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constantS</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.FreqSampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">FreqSampler</code><a class="headerlink" href="#fwdpy.fwdpy.FreqSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a> to track the frequencies of selected mutations over time.</p>
<p>This type is a model of an iterable container.  Return values may be either yielded
or accessed via [i].</p>
<dl class="method">
<dt id="fwdpy.fwdpy.FreqSampler.fetch">
<code class="descname">fetch</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.FreqSampler.fetch" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetch a filtered data set on allele frequency trajectories.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>i</strong> &#8211; The index of the data set to retrieve.</li>
<li><strong>origin_filter</strong> &#8211; (None) A callable to filter on the origin time of mutations.</li>
<li><strong>pos_esize_filter</strong> &#8211; (None) A callable to filter on the position and effect size of a mutation.</li>
<li><strong>freq_filter</strong> &#8211; (None) A callable to filter on the frequency trajectory itself.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Each callable must return True or False.  The callable &#8220;origin_filter&#8221; will receive a single,
non-negative integer for an argument.  &#8220;pos_esize_filter&#8221; will recieve a tuple with two elements,
position and effect size, respectively. Finally, &#8220;freq_filter&#8221; will recieve a list of tuples.  Each
tuple will contain (generation, freq) and will be sorted by generation in ascending order.</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.FreqSampler.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.FreqSampler.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.FreqSampler.to_sql">
<code class="descname">to_sql</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.FreqSampler.to_sql" title="Permalink to this definition">¶</a></dt>
<dd><p>Write output directly to SQLite database files.  Unlike
<a class="reference internal" href="#fwdpy.fwdpy.FreqSampler.fetch" title="fwdpy.fwdpy.FreqSampler.fetch"><code class="xref py py-meth docutils literal"><span class="pre">fetch()</span></code></a>, this function
requires an object of type <a class="reference internal" href="#fwdpy.fwdpy.TrajFilter" title="fwdpy.fwdpy.TrajFilter"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TrajFilter</span></code></a>
to filter out unwanted trajectories.</p>
<p>This function skips the copy of data from C++ to Python, which may make it 
more efficient than using <a class="reference internal" href="#fwdpy.fwdpy.FreqSampler.fetch" title="fwdpy.fwdpy.FreqSampler.fetch"><code class="xref py py-meth docutils literal"><span class="pre">fetch()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>dbname</strong> &#8211; Either the name of a database file (when onedb is True), or the prefix for file names (when onedb is False).</li>
<li><strong>traj_filter</strong> &#8211; (None)  If None, <a class="reference internal" href="#fwdpy.fwdpy.TrajFilter" title="fwdpy.fwdpy.TrajFilter"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TrajFilter</span></code></a> is used, which means all trajectories are written to file.  Otherwise, a custom object is used to filter.</li>
<li><strong>threshold</strong> &#8211; (1,000,000) When onedb is True, this is the number of records to write to the in-memory database before writing to file.</li>
<li><strong>label</strong> &#8211; <ol class="arabic" start="0">
<li>The starting value of the replicate id. When onedb is True, data from different replicates will have a &#8220;rep&#8221; column in the database, with rep goring from threshold to threshold + len(self)-1.</li>
</ol>
</li>
<li><strong>onedb</strong> &#8211; (False)  If False, each trajectory is written to a separate file.  If true, data are first written to in-memory databases and then flushed to disk at time intervals depending on the value of &#8220;threshold&#8221;.</li>
<li><strong>append</strong> &#8211; (False) If false, the output file will be deleted if it exsists.  Otherwise, it will be assumed to be a valid SQLite database and appended to.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The schema of the resulting files can be checked with the sqlite3 command-line tool.</p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.GSLrng">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">GSLrng</code><a class="headerlink" href="#fwdpy.fwdpy.GSLrng" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>A wrapper around a random number generator (rng) 
from the GNU Scientific Library (GSL).</p>
<p>The constructor takes a seed (int) as an argument.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.GammaS">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">GammaS</code><a class="headerlink" href="#fwdpy.fwdpy.GammaS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></p>
<p>Gamma distribution of fitness effects</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>mean: mean of the Gamma</li>
<li>shape: shape of the Gamma</li>
<li>h: the dominance term</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation</p>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>mean</strong> &#8211; the mean selection coefficient</li>
<li><strong>shape</strong> &#8211; the shape parameter of the distribution</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gdist</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GammaS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.35</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.GaussianS">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">GaussianS</code><a class="headerlink" href="#fwdpy.fwdpy.GaussianS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></p>
<p>Gaussian distribution on selection coefficients (effect sizes for sims of quantitative traits)</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>sd: the standard deviation</li>
<li>h: the dominance ter</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>The mean is zero.</p>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation</p>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>mean</strong> &#8211; the mean selection coefficient</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#s N(0,0.1) and co-dominant</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constantS</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GaussianS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.MetaPop">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">MetaPop</code><a class="headerlink" href="#fwdpy.fwdpy.MetaPop" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></p>
<p>Object representing data structures for single-deme simulations.</p>
<p>Users are not expected to construct these on their own.  Rather,
they should be working with <a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">MetaPopVec</span></code></a>.  This type exists as
the output of iterating through a <a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">MetaPopVec</span></code></a>.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.MetaPop.gen">
<code class="descname">gen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPop.gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the generation that the population is currently evolved to</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MetaPop.popsizes">
<code class="descname">popsizes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPop.popsizes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the size of the population</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MetaPop.sane">
<code class="descname">sane</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPop.sane" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes sure that the population is in a sane state.</p>
<p>Internally, this checks that pop.N == pop.diploids.size(),
which it should be if the C++ code behind this all is properly updating
the data structures!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.MetaPopVec">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">MetaPopVec</code><a class="headerlink" href="#fwdpy.fwdpy.MetaPopVec" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></p>
<p>Vector of metapopulation objects.</p>
<p>Internally, the class contains &#8220;smart&#8221; pointers to the underlying C++ types.</p>
<p>The class is iterable, yielding <a class="reference internal" href="#fwdpy.fwdpy.MetaPop" title="fwdpy.fwdpy.MetaPop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPop</span></code></a> objects.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.MetaPopVec.append">
<code class="descname">append</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPopVec.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append &#8216;p&#8217; into this object.</p>
<p>This is done via a serialized copy, meaning that 
this object and p will not share any pointers</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MetaPopVec.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPopVec.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Frees the memory allocated by each simulated population.</p>
<p>Result is an empty object.</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.MetaPopVec.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.MetaPopVec.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MetaPopVec.size">
<code class="descname">size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MetaPopVec.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of populations (size of underlying C++ vector)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.MlocusPop">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">MlocusPop</code><a class="headerlink" href="#fwdpy.fwdpy.MlocusPop" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></p>
<p>Object representing data structures for single-deme, multi-locus/region simulations
based on a mutation type having a single &#8216;s&#8217; and &#8216;h&#8217; term.</p>
<p>Users are not expected to construct these on their own.  Rather,
they should be working with <a class="reference internal" href="#fwdpy.fwdpy.MlocusPopVec" title="fwdpy.fwdpy.MlocusPopVec"><code class="xref py py-class docutils literal"><span class="pre">MlocusPopVec</span></code></a>.  This type exists as
the output of iterating through a <a class="reference internal" href="#fwdpy.fwdpy.MlocusPopVec" title="fwdpy.fwdpy.MlocusPopVec"><code class="xref py py-class docutils literal"><span class="pre">MlocusPopVec</span></code></a>.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPop.gen">
<code class="descname">gen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPop.gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the generation that the population is currently evolved to</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPop.popsize">
<code class="descname">popsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPop.popsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the size of the population</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPop.sane">
<code class="descname">sane</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPop.sane" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes sure that the population is in a sane state.</p>
<p>Internally, this checks that pop.N == pop.diploids.size(),
which it should be if the C++ code behind this all is properly updating
the data structures!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.MlocusPopVec">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">MlocusPopVec</code><a class="headerlink" href="#fwdpy.fwdpy.MlocusPopVec" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></p>
<p>Vector of single-deme objects representing multiple partially-linked regions.</p>
<p>Internally, the class contains &#8220;smart&#8221; pointers to the underlying C++ types.</p>
<p>The class is iterable, yielding <a class="reference internal" href="#fwdpy.fwdpy.MlocusPop" title="fwdpy.fwdpy.MlocusPop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MlocusPop</span></code></a> objects.</p>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for use cases.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPopVec.append">
<code class="descname">append</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPopVec.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append &#8216;p&#8217; into this object.</p>
<p>This is done via a serialized copy, meaning that 
this object and p will not share any pointers</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPopVec.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPopVec.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Frees the memory allocated by each simulated population.</p>
<p>Result is an empty object.</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.MlocusPopVec.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.MlocusPopVec.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.MlocusPopVec.size">
<code class="descname">size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.MlocusPopVec.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of populations (size of underlying C++ vector)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.NothingSampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">NothingSampler</code><a class="headerlink" href="#fwdpy.fwdpy.NothingSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a> that does nothing.</p>
<p>This is useful during evolution to equilibrium or in other situations
where you are not interested in recording.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.NothingSampler.get">
<code class="descname">get</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.NothingSampler.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the data from the sampler.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">None</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.PopSampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">PopSampler</code><a class="headerlink" href="#fwdpy.fwdpy.PopSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a> that takes a sample of size <span class="math">\(n \leq N\)</span> from the population.</p>
<p>This type is a model of an iterable container.  Return values may be either yielded
or accessed via [i].</p>
<dl class="attribute">
<dt id="fwdpy.fwdpy.PopSampler.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.PopSampler.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.PopType">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">PopType</code><a class="headerlink" href="#fwdpy.fwdpy.PopType" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Empty base class for a population object.</p>
<p>Example derived types include <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.singlepop</span></code> and <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.metapop</span></code></p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.PopVec">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">PopVec</code><a class="headerlink" href="#fwdpy.fwdpy.PopVec" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Empty base class for containers of population objects.</p>
<p>Example derived types include <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.SpopVec</span></code></a> and <a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a></p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.QtraitStatsSampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">QtraitStatsSampler</code><a class="headerlink" href="#fwdpy.fwdpy.QtraitStatsSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a> that records various statistics about the population.</p>
<p>This type is a model of an iterable container.  Return values may be either yielded
or accessed via [i].</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This is not useful for the standard fwdpy population.  It only actually records anything meaningful in the qtrait and qtrait_mloc modules.  This will change in a future release.</p>
</div>
<dl class="method">
<dt id="fwdpy.fwdpy.QtraitStatsSampler.get">
<code class="descname">get</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.QtraitStatsSampler.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the data from the sampler.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This returns all data as a list.  It is more RAM-friendly to iterate over the object.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.QtraitStatsSampler.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.QtraitStatsSampler.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.Region">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">Region</code><a class="headerlink" href="#fwdpy.fwdpy.Region" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Representation of a &#8220;region&#8221; in a simulation.</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd>b: the beginning of the region
e: the end of the region
w: the &#8220;weight&#8221; assigned to the region
l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation.</p>
<dl class="docutils">
<dt>This class is extended by:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></li>
</ul>
</dd>
</dl>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#A more &quot;biological&quot; case:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#  The region covers positions 1 through 1,000,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#  and the per-base pair &quot;weight&quot; is 1e-5:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">r</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="mi">1</span><span class="n">e</span><span class="o">-</span><span class="mi">5</span><span class="p">,</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.Spop">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">Spop</code><a class="headerlink" href="#fwdpy.fwdpy.Spop" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></p>
<p>Object representing data structures for single-deme simulations
based on a mutation type having a single &#8216;s&#8217; and &#8216;h&#8217; term.</p>
<p>Users are not expected to construct these on their own.  Rather,
they should be working with <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">SpopVec</span></code></a>.  This type exists as
the output of iterating through a <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">SpopVec</span></code></a>.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.Spop.gen">
<code class="descname">gen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.Spop.gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the generation that the population is currently evolved to</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.Spop.popsize">
<code class="descname">popsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.Spop.popsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the size of the population</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.Spop.sane">
<code class="descname">sane</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.Spop.sane" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes sure that the population is in a sane state.</p>
<p>Internally, this checks that pop.N == pop.diploids.size(),
which it should be if the C++ code behind this all is properly updating
the data structures!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.SpopGenMut">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">SpopGenMut</code><a class="headerlink" href="#fwdpy.fwdpy.SpopGenMut" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></p>
<p>Object representing data structures for single-deme simulations
based on a mutation type having a vector of properties.</p>
<p>Users are not expected to construct these on their own.  Rather,
they should be working with <code class="xref py py-class docutils literal"><span class="pre">SpopGenMutVec</span></code>.  This type exists as
the output of iterating through a <a class="reference internal" href="#fwdpy.fwdpy.SpopGenMut" title="fwdpy.fwdpy.SpopGenMut"><code class="xref py py-class docutils literal"><span class="pre">SpopGenMut</span></code></a>.</p>
<p>..note:: Currently, there are no functions in fwdpy using this type!  See <a class="reference internal" href="#fwdpy.fwdpy.Spop" title="fwdpy.fwdpy.Spop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Spop</span></code></a> instead.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.SpopGenMut.gen">
<code class="descname">gen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopGenMut.gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the generation that the population is currently evolved to</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.SpopGenMut.popsize">
<code class="descname">popsize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopGenMut.popsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the size of the population</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.SpopGenMut.sane">
<code class="descname">sane</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopGenMut.sane" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes sure that the population is in a sane state.</p>
<p>Internally, this checks that pop.N == pop.diploids.size(),
which it should be if the C++ code behind this all is properly updating
the data structures!</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.SpopVec">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">SpopVec</code><a class="headerlink" href="#fwdpy.fwdpy.SpopVec" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></p>
<p>Vector of single-deme objects</p>
<p>Internally, the class contains &#8220;smart&#8221; pointers to the underlying C++ types.</p>
<p>The class is iterable, yielding <a class="reference internal" href="#fwdpy.fwdpy.Spop" title="fwdpy.fwdpy.Spop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Spop</span></code></a> objects.</p>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for use cases.</p>
<dl class="method">
<dt id="fwdpy.fwdpy.SpopVec.append">
<code class="descname">append</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopVec.append" title="Permalink to this definition">¶</a></dt>
<dd><p>Append &#8216;p&#8217; into this object.</p>
<p>This is done via a serialized copy, meaning that 
this object and p will not share any pointers</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.SpopVec.clear">
<code class="descname">clear</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopVec.clear" title="Permalink to this definition">¶</a></dt>
<dd><p>Frees the memory allocated by each simulated population.</p>
<p>Result is an empty object.</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.SpopVec.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.SpopVec.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="fwdpy.fwdpy.SpopVec.size">
<code class="descname">size</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.SpopVec.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns number of populations (size of underlying C++ vector)</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.Sregion">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">Sregion</code><a class="headerlink" href="#fwdpy.fwdpy.Sregion" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Region" title="fwdpy.fwdpy.Region"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Region</span></code></a></p>
<p>Representation of a &#8220;region&#8221; in a simulation with a dominance term.</p>
<p>This class is the base class for a general set of objects representing distributions of fitness effects.</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>h: the dominance term</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This class cannot be used directly to parameterize a simulation.  Rather, you must used
a derived type that specifies a distribution of fitness effects.  These types include:
<a class="reference internal" href="#fwdpy.fwdpy.ConstantS" title="fwdpy.fwdpy.ConstantS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.ConstantS</span></code></a>,
<a class="reference internal" href="#fwdpy.fwdpy.UniformS" title="fwdpy.fwdpy.UniformS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.UniformS</span></code></a>,
<a class="reference internal" href="#fwdpy.fwdpy.ExpS" title="fwdpy.fwdpy.ExpS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.ExpS</span></code></a>,
<a class="reference internal" href="#fwdpy.fwdpy.GammaS" title="fwdpy.fwdpy.GammaS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.GammaS</span></code></a>, and 
<a class="reference internal" href="#fwdpy.fwdpy.GaussianS" title="fwdpy.fwdpy.GaussianS"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.GaussianS</span></code></a></p>
</div>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Examples for models where the 3 genotype fitnesses are</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#1, 1+sh, and 1+2s, respectively</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">recessive</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Sregion</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">additive</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Sregion</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dominant</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">Sregion</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">2.0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.TemporalSampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">TemporalSampler</code><a class="headerlink" href="#fwdpy.fwdpy.TemporalSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base class representing containers of functions
to be applied to simluated populations at regular intervals.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.TrajExistedPast">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">TrajExistedPast</code><a class="headerlink" href="#fwdpy.fwdpy.TrajExistedPast" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TrajFilter" title="fwdpy.fwdpy.TrajFilter"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TrajFilter</span></code></a></p>
<p>A type of <a class="reference internal" href="#fwdpy.fwdpy.TrajFilter" title="fwdpy.fwdpy.TrajFilter"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TrajFilter</span></code></a>.  Tests that a mutation existed
in the simulation past a certain generation.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.TrajFilter">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">TrajFilter</code><a class="headerlink" href="#fwdpy.fwdpy.TrajFilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base class for filtering trajectories.</p>
<p>An instance of this object allows all trajectories to be written to file.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">See <a class="reference internal" href="#fwdpy.fwdpy.FreqSampler.to_sql" title="fwdpy.fwdpy.FreqSampler.to_sql"><code class="xref py py-meth docutils literal"><span class="pre">to_sql()</span></code></a></p>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.UniformS">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">UniformS</code><a class="headerlink" href="#fwdpy.fwdpy.UniformS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a></p>
<p>Uniform distribution on selection coefficients</p>
<dl class="docutils">
<dt>Attributes:</dt>
<dd><ul class="first last simple">
<li>b: the beginning of the region</li>
<li>e: the end of the region</li>
<li>w: the &#8220;weight&#8221; assigned to the region</li>
<li>lo: the lower bound on s</li>
<li>hi: the upper bound on s</li>
<li>h: the dominance term</li>
<li>l: A label assigned to the region.  Labels must be integers, and can be used to &#8216;tag&#8217; mutations arising in different regions.</li>
</ul>
</dd>
</dl>
<p>See <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">evolve_regions()</span></code></a> for how this class may be used to parameterize a simulation</p>
<p>Constructor</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>beg</strong> &#8211; the beginning of the region</li>
<li><strong>end</strong> &#8211; the end of the region</li>
<li><strong>weight</strong> &#8211; the weight to assign</li>
<li><strong>lo</strong> &#8211; lower bound on s</li>
<li><strong>hi</strong> &#8211; upper bound on s</li>
<li><strong>h</strong> &#8211; the dominance</li>
<li><strong>coupled</strong> &#8211; if True, the weight is converted to (end-beg)*weight</li>
<li><strong>label</strong> &#8211; Not relevant to recombining regions.  Otherwise, this value will be used to take mutations from this region.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>When coupled is True, the &#8220;weight&#8221; may be interpreted as a &#8220;per base pair&#8221;
(or per unit, generally speaking) term.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1">#A simple case</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#s is uniform on [0,-1]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">constantS</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">UniformS</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpy.VASampler">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpy.</code><code class="descname">VASampler</code><a class="headerlink" href="#fwdpy.fwdpy.VASampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a> that estimates the relationship between mutation frequency and total additive
genetic variance.</p>
<p>This type is a model of an iterable container.  Return values may be either yielded
or accessed via [i].</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This is not useful for the standard fwdpy population.  It only actually records anything meaningful in the qtrait and qtrait_mloc modules.  This will change in a future release.</p>
</div>
<dl class="method">
<dt id="fwdpy.fwdpy.VASampler.get">
<code class="descname">get</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.VASampler.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve the data from the sampler.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This returns all data as a list.  It is more RAM-friendly to iterate over the object.</p>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.fwdpy.VASampler.next">
<code class="descname">next</code><a class="headerlink" href="#fwdpy.fwdpy.VASampler.next" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.add_mutation">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">add_mutation</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.add_mutation" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a mutation to a population at a specific position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>p</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a>.</li>
<li><strong>indlist</strong> &#8211; A list of individuals to whom the mutation will be added.  Must be unsigned indexes [0,N).</li>
<li><strong>clist</strong> &#8211; A list of chromosomes to which to add the new mutation in each individual. 0 = chrom1, 1 = chrom2, 2 = both.</li>
<li><strong>pos</strong> &#8211; The position of the new mutation</li>
<li><strong>s</strong> &#8211; The selection coefficient/effect size of the new mutation</li>
<li><strong>h</strong> &#8211; The dominance coefficient of the new mutation</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Only <a class="reference internal" href="#fwdpy.fwdpy.Spop" title="fwdpy.fwdpy.Spop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Spop</span></code></a> is currently supported</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">RuntimeError will be thrown if pos already exists in p or if any element in clist and/or indlist is out of range.</p>
</div>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span> <span class="k">as</span> <span class="nn">fp</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#init a monomorphic pop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">SpopVec</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">add_mutation</span><span class="p">(</span><span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">20</span><span class="p">],[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.0</span><span class="p">,</span><span class="mf">1.0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#x will have value 0 b/c it is first mutation entered into pop</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fp</span><span class="o">.</span><span class="n">view_mutations</span><span class="p">(</span><span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="go">[{&#39;g&#39;: 0, &#39;h&#39;: 1.0, &#39;pos&#39;: 0.01, &#39;n&#39;: 4, &#39;s&#39;: 0.0, &#39;ftime&#39;: 4294967295, &#39;label&#39;: 0, &#39;neutral&#39;: True}]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.apply_sampler">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">apply_sampler</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.apply_sampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a temporal sampler to a container of populations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pops</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></li>
<li><strong>sampler</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Nothing</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.apply_sampler_single">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">apply_sampler_single</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.apply_sampler_single" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply a temporal sampler to an indivudal <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pop</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></li>
<li><strong>sampler</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The use case for this function is applying very expensive temporal samplers
at the end of a simulation.  It is assumed that len(sampler)==1.</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.check_popdata">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">check_popdata</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.check_popdata" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply fwdpp&#8217;s debugging functions to population containers.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>p</strong> &#8211; A object of type <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.poptype</span></code> or <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.poptype</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">Dictionary with return values (True or False) Any false values reflect a critical data inconsistency, and mean an exception should be raised.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.cite">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">cite</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.cite" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns how to cite this package</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.copypop">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">copypop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.copypop" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy a population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pop</strong> &#8211; the population to copy.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">A <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> of the same (derived) type as pop</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The function is implemented in terms of the serialization routines in the fwdpy.fwdpyio module.  The return value can be evolved and not affect the input value.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.copypops">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">copypops</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.copypops" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy a population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pops</strong> &#8211; the list of population to copy.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">A <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a> of the same (derived) type as pop</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The function is implemented in terms of the serialization routines in the fwdpy.fwdpyio module.  The return value can be evolved and not affect the input value.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.diploid_mloc_view_to_sample">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">diploid_mloc_view_to_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.diploid_mloc_view_to_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a &#8220;view&#8221; of diploids into the data types returned from <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>view</strong> &#8211; An object returned by <a class="reference internal" href="#fwdpy.fwdpy.view_diploids" title="fwdpy.fwdpy.view_diploids"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.view_diploids()</span></code></a> from a multilocus population.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A dict.  See Note</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The return value is a dict containing two lists of tuples (&#8216;neutral&#8217;, and &#8216;selected&#8217;)  These lists are the same format 
as objects returned by <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a>.  The dict also contains &#8216;neutral_info&#8217; and &#8216;selected_info&#8217;, which are
pandas.DataFrame objects with the information about the mutations present in the other two lists.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.diploid_traits">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">diploid_traits</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.diploid_traits" title="Permalink to this definition">¶</a></dt>
<dd><p>Return genetic value (g), environmental value (e), and fitness (w) for all diploids.</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.diploid_view_to_sample">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">diploid_view_to_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.diploid_view_to_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a &#8220;view&#8221; of diploids into the data types returned from <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>view</strong> &#8211; An object returned by <a class="reference internal" href="#fwdpy.fwdpy.view_diploids" title="fwdpy.fwdpy.view_diploids"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.view_diploids()</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A dict.  See Note</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The return value is a dict containing two lists of tuples (&#8216;neutral&#8217;, and &#8216;selected&#8217;)  These lists are the same format 
as objects returned by <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a>.  The dict also contains &#8216;neutral_info&#8217; and &#8216;selected_info&#8217;, which are
pandas.DataFrame objects with the information about the mutations present in the other two lists.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.evolve_regions">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">evolve_regions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.evolve_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve a region with variable mutation, fitness effects, and recombination rates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>npops</strong> &#8211; The number of populations to simulate.  This is equal to the number of threads that will be used!</li>
<li><strong>N</strong> &#8211; The diploid population size to simulate</li>
<li><strong>nlist</strong> &#8211; An array view of a NumPy array.  This represents the population sizes over time.  The length of this view is the length of the simulation in generations. The view must be of an array of 32 bit, unsigned integers (see example).</li>
<li><strong>mu_neutral</strong> &#8211; The mutation rate to variants not affecting fitness (&#8220;neutral&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>mu_selected</strong> &#8211; The mutation rate to variants affecting fitness (&#8220;selected&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>recrate</strong> &#8211; The recombination rate in the regions (per diploid, per generation)</li>
<li><strong>nregions</strong> &#8211; A list specifying where neutral mutations occur</li>
<li><strong>sregions</strong> &#8211; A list specifying where selected mutations occur</li>
<li><strong>recregions</strong> &#8211; A list specifying how the genetic map varies along the region</li>
<li><strong>f</strong> &#8211; The selfing probabilty</li>
<li><strong>scaling</strong> &#8211; For a single mutation, fitness is calculated as 1, 1+sh, and 1+scaling*s for genotypes AA, Aa, and aa, respectively.</li>
<li><strong>fitness</strong> &#8211; The fitness model.  Must be either &#8220;multiplicative&#8221; or &#8220;additive&#8221;.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last">RuntimeError if parameters do not pass checks</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Our &quot;neutral&quot; regions will be from positions [0,1) and [2,3).</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#The regions will have equal weights, and thus will each get</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#1/2 of newly-arising, neutral mutations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Our &quot;selected&quot; mutations will have positions on the continuous</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#interval [1,2).  There will be two classes of such mutations,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#each with exponentially-distrubted selection coefficients.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#The first class will have a mean of s = -0.1 (deleterious), and the second</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#will have a mean of 1e-3 (adaptive).  The former will be 100x more common than</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#the latter, as the weights are 1 and 0.01, respectively.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Recombination will be uniform along the interval [0,3).</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#popsizes are NumPy arrays of 32bit unsigned integers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Initial pop size will be N = 1,000, and</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#the type must be uint32 (32 bit unsigned integer)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#The simulation will be for 10*N generations,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#so we replicate that value in the array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Simulate 1 deme under this model.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#The total neutral mutation rate is 1e-3,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#which is also the recombination rate.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#The total mutation rate to selected variants is 0.1*(the neutral mutation rate).</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.evolve_regions_more">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">evolve_regions_more</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.evolve_regions_more" title="Permalink to this definition">¶</a></dt>
<dd><p>Continue to evolve a region with variable mutation, fitness effects, and recombination rates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>pops</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">SpopVec</span></code></a></li>
<li><strong>nlist</strong> &#8211; An array view of a NumPy array.  This represents the population sizes over time.  The length of this view is the length of the simulation in generations. The view must be of an array of 32 bit, unsigned integers (see example).</li>
<li><strong>mu_neutral</strong> &#8211; The mutation rate to variants not affecting fitness (&#8220;neutral&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>mu_selected</strong> &#8211; The mutation rate to variants affecting fitness (&#8220;selected&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>recrate</strong> &#8211; The recombination rate in the regions (per diploid, per generation)</li>
<li><strong>nregions</strong> &#8211; A list specifying where neutral mutations occur</li>
<li><strong>sregions</strong> &#8211; A list specifying where selected mutations occur</li>
<li><strong>recregions</strong> &#8211; A list specifying how the genetic map varies along the region</li>
<li><strong>f</strong> &#8211; The selfing probabilty</li>
<li><strong>scaling</strong> &#8211; For a single mutation, fitness is calculated as 1, 1+sh, and 1+scaling*s for genotypes AA, Aa, and aa, respectively.</li>
<li><strong>fitness</strong> &#8211; The fitness model.  Must be either &#8220;multiplicative&#8221; or &#8220;additive&#8221;.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last">RuntimeError if parameters do not pass checks</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># See docstring for fwdpy.evolve_regions for the gory details</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Evolve for 5N generations initially</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">5000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Evolve for another 5N generations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions_more</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">pops</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.evolve_regions_sampler">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">evolve_regions_sampler</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.evolve_regions_sampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve a single population under standard population genetic fitness models and apply a &#8220;sampler&#8221; at regular intervals.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>pops</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">SpopVec</span></code></a></li>
<li><strong>slist</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">TemporalSampler</span></code></a>.</li>
<li><strong>nlist</strong> &#8211; An array view of a NumPy array.  This represents the population sizes over time.  The length of this view is the length of the simulation in generations. The view must be of an array of 32 bit, unsigned integers (see example).</li>
<li><strong>mu_neutral</strong> &#8211; The mutation rate to variants not affecting fitness (&#8220;neutral&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>mu_selected</strong> &#8211; The mutation rate to variants affecting fitness (&#8220;selected&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>recrate</strong> &#8211; The recombination rate in the regions (per diploid, per generation)</li>
<li><strong>nregions</strong> &#8211; A list specifying where neutral mutations occur</li>
<li><strong>sregions</strong> &#8211; A list specifying where selected mutations occur</li>
<li><strong>recregions</strong> &#8211; A list specifying how the genetic map varies along the region</li>
<li><strong>sample</strong> &#8211; Apply the temporal sample every &#8216;sample&#8217; generations during the simulation.</li>
<li><strong>f</strong> &#8211; The selfing probabilty</li>
<li><strong>scaling</strong> &#8211; For a single mutation, fitness is calculated as 1, 1+sh, and 1+scaling*s for genotypes AA, Aa, and aa, respectively.</li>
<li><strong>fitness</strong> &#8211; The fitness model.  Must be either &#8220;multiplicative&#8221; or &#8220;additive&#8221;.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.evolve_regions_sampler_fitness">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">evolve_regions_sampler_fitness</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.evolve_regions_sampler_fitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve a single population under arbitrary fitness models and apply a &#8220;sampler&#8221; at regular intervals.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>pops</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.SpopVec" title="fwdpy.fwdpy.SpopVec"><code class="xref py py-class docutils literal"><span class="pre">SpopVec</span></code></a></li>
<li><strong>slist</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">TemporalSampler</span></code></a>.</li>
<li><strong>fitness_function</strong> &#8211; A <a class="reference internal" href="#fwdpy.fitness.SpopFitness" title="fwdpy.fitness.SpopFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.SpopFitness</span></code></a></li>
<li><strong>nlist</strong> &#8211; An array view of a NumPy array.  This represents the population sizes over time.  The length of this view is the length of the simulation in generations. The view must be of an array of 32 bit, unsigned integers (see example).</li>
<li><strong>mu_neutral</strong> &#8211; The mutation rate to variants not affecting fitness (&#8220;neutral&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>mu_selected</strong> &#8211; The mutation rate to variants affecting fitness (&#8220;selected&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>recrate</strong> &#8211; The recombination rate in the regions (per diploid, per generation)</li>
<li><strong>nregions</strong> &#8211; A list specifying where neutral mutations occur</li>
<li><strong>sregions</strong> &#8211; A list specifying where selected mutations occur</li>
<li><strong>recregions</strong> &#8211; A list specifying how the genetic map varies along the region</li>
<li><strong>sample</strong> &#8211; Apply the temporal sampler every &#8216;sample&#8217; generations during the simulation. 0 means it will never get applied, which may or may not be what you want.</li>
<li><strong>f</strong> &#8211; The selfing probabilty</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.freqfilter">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">freqfilter</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.freqfilter" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove low-frequency variants from a sample.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> &#8211; a sample from a population.  For example, the return value of <a class="reference internal" href="#fwdpy.fwdpy.ms_sample" title="fwdpy.fwdpy.ms_sample"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.ms_sample()</span></code></a> or <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a></li>
<li><strong>minfreq</strong> &#8211; Remove all sites with frequency &lt; minfreq</li>
<li><strong>derived</strong> &#8211; if True, filter on derived allele frequency.  If False, filter on minor allele frequency.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s2</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">freqfilter</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="mf">0.2</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.get_includes">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">get_includes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.get_includes" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns absolute path to location of fwdpy headers</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.get_sample_details">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">get_sample_details</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.get_sample_details" title="Permalink to this definition">¶</a></dt>
<dd><p>Get additional details for population samples</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ms_samples</strong> &#8211; A list returned by <a class="reference internal" href="#fwdpy.fwdpy.ms_sample" title="fwdpy.fwdpy.ms_sample"><code class="xref py py-func docutils literal"><span class="pre">ms_sample()</span></code></a></li>
<li><strong>pops</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">PopType</span></code></a></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Params locusID:</th><td class="field-body"><p class="first">(None) A numerical label for a locus.  Only used/relevant if pop is <a class="reference internal" href="#fwdpy.fwdpy.MlocusPop" title="fwdpy.fwdpy.MlocusPop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MlocusPop</span></code></a></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A dictionary of data for each mutation in the sample.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">details</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">get_sample_details</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">s</span><span class="p">,</span><span class="n">pop</span><span class="p">)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.get_samples">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">get_samples</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.get_samples" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a sample from a set of simulated populations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>pop</strong> &#8211; An object inheriting from <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">PopType</span></code></a></li>
<li><strong>nsam</strong> &#8211; The sample size to take.</li>
<li><strong>removeFixed</strong> &#8211; if True, only polymorphic sites are retained</li>
<li><strong>deme</strong> &#8211; Optional.  If &#8216;pop&#8217; is a <a class="reference internal" href="#fwdpy.fwdpy.MetaPop" title="fwdpy.fwdpy.MetaPop"><code class="xref py py-class docutils literal"><span class="pre">MetaPop</span></code></a>, deme is required and represents the sub-population to sample.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A list. Element 0 is neutral mutations, and element 1 is selected mutations.  Within each list is a tuple of size 2.  The first element is the mutation position.  The second element is the genotype for each of the &#8216;nsam&#8217; chromosomes.  Genotypes are coded as 0 = the ancestral state and 1 = the derived state.  For each site, each pair of genotypes constitutes a single diploid.  In other words, for nsam = 50, the data will represent the complete haplotypes of 25 diploids.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raise:</th><td class="field-body"><p class="first last">IndexError if &#8216;deme&#8217; is out of range and pop is a <a class="reference internal" href="#fwdpy.fwdpy.MetaPop" title="fwdpy.fwdpy.MetaPop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPop</span></code></a></p>
</td>
</tr>
</tbody>
</table>
<p>Please note that if you desire an odd &#8216;nsam&#8217;, you should input nsam+2 and randomly remove one haplotype to obtain your desired sample size.  This is due to an issue with how we are sampling chromosomes from the population.</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">get_samples</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.getfreq">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">getfreq</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.getfreq" title="Permalink to this definition">¶</a></dt>
<dd><p>Get mutation frequencies</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>site</strong> &#8211; A tuple. See example</li>
<li><strong>derived</strong> &#8211; If True, report derived allele frequency (DAF).  If False, return minor allele freqency (MAF)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Do <strong>not</strong> use this function to calculate <span class="math">\(\pi\)</span> (a.k.a. <span class="math">\(\hat\theta_\pi\)</span>, a.k.a. &#8220;sum of site heterozygosity&#8221;).
<span class="math">\(\pi\)</span> for a <strong>sample</strong> is not <span class="math">\(2\sum_ip_iq_i\)</span>. because the sample is <em>finite</em>.
In general, it will be more convenient to call <a class="reference internal" href="#fwdpy.fwdpy.getfreqs" title="fwdpy.fwdpy.getfreqs"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.getfreqs()</span></code></a> on a list of tuples.</p>
</div>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#DAF = 7/10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#MAF = 3/10</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="s1">&#39;0111111100&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">round</span><span class="p">(</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">getfreq</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="kc">True</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
<span class="go">0.7</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">round</span><span class="p">(</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">getfreq</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="kc">False</span><span class="p">),</span><span class="mi">3</span><span class="p">)</span>
<span class="go">0.3</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.getfreqs">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">getfreqs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.getfreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#fwdpy.fwdpy.getfreq" title="fwdpy.fwdpy.getfreq"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.getfreq()</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>sample</strong> &#8211; a sample from a population.  For example, the return value of <a class="reference internal" href="#fwdpy.fwdpy.ms_sample" title="fwdpy.fwdpy.ms_sample"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.ms_sample()</span></code></a> or <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a></li>
<li><strong>derived</strong> &#8211; If True, report derived allele frequency (DAF).  If False, return minor allele freqency (MAF).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Do <strong>not</strong> use this function to calculate <span class="math">\(\pi\)</span> (a.k.a. <span class="math">\(\hat\theta_\pi\)</span>, a.k.a. &#8220;sum of site heterozygosity&#8221;).
<span class="math">\(\pi\)</span> for a <strong>sample</strong> is not <span class="math">\(2\sum_ip_iq_i\)</span>. because the sample is <em>finite</em>.</p>
</div>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">freqs</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">getfreqs</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.ms_sample">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">ms_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.ms_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Take a sample from a set of simulated populations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>rng</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.GSLrng" title="fwdpy.fwdpy.GSLrng"><code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></a></li>
<li><strong>pops</strong> &#8211; An object inheriting from <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">PopType</span></code></a></li>
<li><strong>nsam</strong> &#8211; List of sample sizes (no. chromosomes) to sample.</li>
<li><strong>removeFixed</strong> &#8211; if True, only polymorphic sites are retained</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>Note: nsam will likely be changed to a list soon, to accomodate multi-deme simulations</p>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.nderived">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">nderived</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.nderived" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#fwdpy.fwdpy.nderived" title="fwdpy.fwdpy.nderived"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.nderived()</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sample</strong> &#8211; a sample from a population.  For example, the return value of <a class="reference internal" href="#fwdpy.fwdpy.ms_sample" title="fwdpy.fwdpy.ms_sample"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.ms_sample()</span></code></a> or <a class="reference internal" href="#fwdpy.fwdpy.get_samples" title="fwdpy.fwdpy.get_samples"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.get_samples()</span></code></a></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span><span class="o">,</span><span class="nn">array</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nd</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">nderived</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.nderived_site">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">nderived_site</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.nderived_site" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the number of derived mutations at a site.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>site</strong> &#8211; A tuple.  See example</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">In general, it will be more convenient to call <a class="reference internal" href="#fwdpy.fwdpy.nderived" title="fwdpy.fwdpy.nderived"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.nderived()</span></code></a> on a list of tuples.</p>
</div>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Create a site at position 0.1 with the</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#genotypes as the second element. 0/1 = ancestral/derived</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="s1">&#39;01100111&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fwdpy</span><span class="o">.</span><span class="n">nderived_site</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#Process simulation results:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">array</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="s1">&#39;I&#39;</span><span class="p">,[</span><span class="mi">1000</span><span class="p">]</span><span class="o">*</span><span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)],[],[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ms_sample</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="n">i</span><span class="p">,</span><span class="mi">10</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pop</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span> <span class="n">ndi</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">nderived_site</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.pkg_version">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">pkg_version</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.pkg_version" title="Permalink to this definition">¶</a></dt>
<dd><p>Return version numbers of this package</p>
<p>This function is very handy when reporting bugs!</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.view_diploids">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">view_diploids</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.view_diploids" title="Permalink to this definition">¶</a></dt>
<dd><p>Get detailed list of a set of diploids in the population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>p</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> or a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></li>
<li><strong>deme</strong> &#8211; if p is a :class`fwdpy.fwdpy.MetaPop`, deme is the index of the deme to sample</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a list of dictionaries.  See Note.</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dips</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">view_diploids</span><span class="p">(</span><span class="n">i</span><span class="p">,[</span><span class="mi">0</span><span class="p">,</span><span class="mi">101</span><span class="p">,</span><span class="mi">201</span><span class="p">,</span><span class="mi">301</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a> currently not supported</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.view_diploids_pd">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">view_diploids_pd</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.view_diploids_pd" title="Permalink to this definition">¶</a></dt>
<dd><p>Get detailed list of a set of diploids in the population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>p</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> or a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></li>
<li><strong>deme</strong> &#8211; if p is a :class`fwdpy.fwdpy.MetaPop`, deme is the index of the deme to sample</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pandas.DataFrame or a list of such objects</p>
</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a> is not yet supported</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.view_fixations">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">view_fixations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.view_fixations" title="Permalink to this definition">¶</a></dt>
<dd><p>Return information on fixed variants</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>p</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> or a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A list of tuples. The first element is fixation time, and the second is a dict containing data about the mutation.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">You may need to call <a class="reference internal" href="#fwdpy.fwdpy.view_mutations" title="fwdpy.fwdpy.view_mutations"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.view_mutations()</span></code></a> to view all types of fixations, depending on the type of simulation you are running.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.view_gametes">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">view_gametes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.view_gametes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get detailed list of all gametes in the population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>p</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> or a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></li>
<li><strong>deme</strong> &#8211; If p is a <a class="reference internal" href="#fwdpy.fwdpy.MetaPop" title="fwdpy.fwdpy.MetaPop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPop</span></code></a>, deme is the index of the deme to view</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">a list of dictionaries.  See note.</p>
</td>
</tr>
</tbody>
</table>
<p>Example for a single deme:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dips</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">view_gametes</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a> currently not supported</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpy.view_mutations">
<code class="descclassname">fwdpy.fwdpy.</code><code class="descname">view_mutations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpy.view_mutations" title="Permalink to this definition">¶</a></dt>
<dd><p>Get detailed list of all mutations in the population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>p</strong> &#8211; a <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a> or a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">a list of dictionaries.  See Note.</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#muts[0] will be a list and muts[0][0] will be a dict</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">muts</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">view_mutations</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last"><a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a> currently not supported</p>
</div>
</dd></dl>

<span class="target" id="module-fwdpy.fitness"></span><dl class="class">
<dt id="fwdpy.fitness.MlocusAdditive">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">MlocusAdditive</code><a class="headerlink" href="#fwdpy.fitness.MlocusAdditive" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.MlocusFitness" title="fwdpy.fitness.MlocusFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.MlocusFitness</span></code></a></p>
<p>Additive fitness for a multi-region model.</p>
<p>Fitness is additive with dominance within regions,
then strictly additive across regions.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fitness.MlocusFitness">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">MlocusFitness</code><a class="headerlink" href="#fwdpy.fitness.MlocusFitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base object for multi-locus fitness functions</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fitness.MlocusMult">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">MlocusMult</code><a class="headerlink" href="#fwdpy.fitness.MlocusMult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.MlocusFitness" title="fwdpy.fitness.MlocusFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.MlocusFitness</span></code></a></p>
<p>Multiplicative fitness for a multi-region model.</p>
<p>Fitness is multiplicative with dominance within regions,
then strictly additive across regions.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fitness.SpopAdditive">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">SpopAdditive</code><a class="headerlink" href="#fwdpy.fitness.SpopAdditive" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.SpopFitness" title="fwdpy.fitness.SpopFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.SpopFitness</span></code></a></p>
<p>Additive fitness model for a single deme.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fitness.SpopFitness">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">SpopFitness</code><a class="headerlink" href="#fwdpy.fitness.SpopFitness" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Base object for single-deme fitness functions</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fitness.SpopMult">
<em class="property">class </em><code class="descclassname">fwdpy.fitness.</code><code class="descname">SpopMult</code><a class="headerlink" href="#fwdpy.fitness.SpopMult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.SpopFitness" title="fwdpy.fitness.SpopFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.SpopFitness</span></code></a></p>
<p>Multiplicative fitness model for a single deme.</p>
</dd></dl>

<span class="target" id="module-fwdpy.demography"></span><span class="target" id="module-fwdpy.matrix"></span><dl class="class">
<dt id="fwdpy.matrix.DataMatrix">
<em class="property">class </em><code class="descclassname">fwdpy.matrix.</code><code class="descname">DataMatrix</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Matrix representation of a sample.
All array types are array.array.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.0.4.</span></p>
</div>
<dl class="method">
<dt id="fwdpy.matrix.DataMatrix.matrix">
<code class="descname">matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.DataMatrix.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns matrix of neutral and selected sites.  If either matrix is empty, the other
is returned.  Otherwise, the neutral and selected matrix are appended (neutral columns
first, then selected).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dtype</strong> &#8211; (numpy.dtype(&#8216;i1&#8217;)) Data type for matrix.  Default is 8-bit integer.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">numpy.array or None if the matrix is empty</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.neutral">
<code class="descname">neutral</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.neutral" title="Permalink to this definition">¶</a></dt>
<dd><p>Data for neutral positions</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.neutral_popfreq">
<code class="descname">neutral_popfreq</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.neutral_popfreq" title="Permalink to this definition">¶</a></dt>
<dd><p>Neutral frequencies in simulation</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.neutral_positions">
<code class="descname">neutral_positions</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.neutral_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>Neutral positions</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.matrix.DataMatrix.nmatrix">
<code class="descname">nmatrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.DataMatrix.nmatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns matrix of neutral sites</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dtype</strong> &#8211; (numpy.dtype(&#8216;i1&#8217;)) Data type for matrix.  Default is 8-bit integer.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">numpy.array or None if the matrix is empty</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.nn">
<code class="descname">nn</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.nn" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of neutral mutations</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.nrow">
<code class="descname">nrow</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.nrow" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of rows</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.ns">
<code class="descname">ns</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.ns" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of selected nutations</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.selected">
<code class="descname">selected</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.selected" title="Permalink to this definition">¶</a></dt>
<dd><p>Data for selected positions</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.selected_popfreq">
<code class="descname">selected_popfreq</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.selected_popfreq" title="Permalink to this definition">¶</a></dt>
<dd><p>selected frequencies in simulation</p>
</dd></dl>

<dl class="attribute">
<dt id="fwdpy.matrix.DataMatrix.selected_positions">
<code class="descname">selected_positions</code><a class="headerlink" href="#fwdpy.matrix.DataMatrix.selected_positions" title="Permalink to this definition">¶</a></dt>
<dd><p>selected positions</p>
</dd></dl>

<dl class="method">
<dt id="fwdpy.matrix.DataMatrix.smatrix">
<code class="descname">smatrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.DataMatrix.smatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns matrix of selected sites</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>dtype</strong> &#8211; (numpy.dtype(&#8216;i1&#8217;)) Data type for matrix.  Default is 8-bit integer.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">numpy.array or None if the matrix is empty</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fwdpy.matrix.GenotypeMatrix">
<em class="property">class </em><code class="descclassname">fwdpy.matrix.</code><code class="descname">GenotypeMatrix</code><a class="headerlink" href="#fwdpy.matrix.GenotypeMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.matrix.DataMatrix" title="fwdpy.matrix.DataMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.DataMatrix</span></code></a></p>
<p>Interface is identical to <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.DataMatrix</span></code></p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.matrix.HaplotypeMatrix">
<em class="property">class </em><code class="descclassname">fwdpy.matrix.</code><code class="descname">HaplotypeMatrix</code><a class="headerlink" href="#fwdpy.matrix.HaplotypeMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.matrix.DataMatrix" title="fwdpy.matrix.DataMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.DataMatrix</span></code></a></p>
<p>An extension of <code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.DataMatrix</span></code>
that allows for conversion to data formats compatible
wiht pylibseq/libsequence.</p>
<dl class="method">
<dt id="fwdpy.matrix.HaplotypeMatrix.as_sample">
<code class="descname">as_sample</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.HaplotypeMatrix.as_sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of the data in a form 
usable in pylibseq.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A dictionary with elements &#8216;neutral&#8217; and &#8216;selected&#8217;</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">dict</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fwdpy.matrix.genotype_matrix">
<code class="descclassname">fwdpy.matrix.</code><code class="descname">genotype_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.genotype_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a genotype matrix stored in a <a class="reference internal" href="#fwdpy.matrix.GenotypeMatrix" title="fwdpy.matrix.GenotypeMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.GenotypeMatrix</span></code></a></p>
<p>All arguments are identical to <a class="reference internal" href="#fwdpy.matrix.get_mutation_keys" title="fwdpy.matrix.get_mutation_keys"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.matrix.get_mutation_keys()</span></code></a> with the exception of:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>keys</strong> &#8211; (None) Pre-calculated mutation keys.  Must be a tuple of lists of positive integers. Exceptions will be raised if keys are invalid or out of range.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#fwdpy.matrix.GenotypeMatrix" title="fwdpy.matrix.GenotypeMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.GenotypeMatrix</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.matrix.get_mutation_keys">
<code class="descclassname">fwdpy.matrix.</code><code class="descname">get_mutation_keys</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.get_mutation_keys" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate keys to the set of mutations present in the sample defined by a set of individuals.
The keys may be used to generate a <a class="reference internal" href="#fwdpy.matrix.DataMatrix" title="fwdpy.matrix.DataMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.DataMatrix</span></code></a> via a call to 
<a class="reference internal" href="#fwdpy.matrix.haplotype_matrix" title="fwdpy.matrix.haplotype_matrix"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.matrix.haplotype_matrix()</span></code></a> and/or <a class="reference internal" href="#fwdpy.matrix.genotype_matrix" title="fwdpy.matrix.genotype_matrix"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.matrix.genotype_matrix()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>pop</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></li>
<li><strong>individuals</strong> &#8211; A list of indexes of individuals in the sample</li>
<li><strong>include_neutral</strong> &#8211; (True) If True, generate keys for neutral mutations</li>
<li><strong>include_selected</strong> &#8211; (True) If True, generate keys for selected mutations</li>
<li><strong>remove_fixed</strong> &#8211; (False) If True, prune the return value of keys to mutations that are fixed in the sample</li>
<li><strong>sort</strong> &#8211; (True) If True, keys are sorted according to mutation positions</li>
<li><strong>min_sample_daf</strong> &#8211; (None) If not None, then it must be a float (0 &lt;= min_sample_daf &lt; 1) and only mutations with derived allele frequency <strong>in the sample</strong> <span class="math">\(q \geq  \mathrm{min_sample_daf}\)</span> will be kept.</li>
<li><strong>min_pop_daf</strong> &#8211; (None) If not None, then it must be a float (0 &lt;= min_pop_daf &lt; 1) and only mutations with derived allele frequency <strong>in the (meta-)population</strong> <span class="math">\(q \geq  \mathrm{min_pop_daf}\)</span> will be kept.</li>
<li><strong>deme</strong> &#8211; If pop is a metapopulation, this is the index of the deme.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Keys to neutral and selected mutations stored in separate lists.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">tuple of two lists</p>
</td>
</tr>
</tbody>
</table>
<div class="versionadded">
<p><span class="versionmodified">New in version 0.0.4.</span></p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">For meta-populations, min_pop_daf applies to the entire meta-population, and not the specific deme.
Other types of key sorting must be done by the user.  This is currently difficult without writing
a custom Cython plugin based on fwdpy.  This will be improved in the future with a new &#8220;views&#8221; functionality.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.matrix.haplotype_matrix">
<code class="descclassname">fwdpy.matrix.</code><code class="descname">haplotype_matrix</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.matrix.haplotype_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate a haplotype matrix stored in a <a class="reference internal" href="#fwdpy.matrix.HaplotypeMatrix" title="fwdpy.matrix.HaplotypeMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.HaplotypeMatrix</span></code></a></p>
<p>All arguments are identical to <a class="reference internal" href="#fwdpy.matrix.get_mutation_keys" title="fwdpy.matrix.get_mutation_keys"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.matrix.get_mutation_keys()</span></code></a> with the exception of:</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>keys</strong> &#8211; (None) Pre-calculated mutation keys.  Must be a tuple of lists of positive integers. Exceptions will be raised if keys are invalid or out of range.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#fwdpy.matrix.HaplotypeMatrix" title="fwdpy.matrix.HaplotypeMatrix"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.matrix.HaplotypeMatrix</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<span class="target" id="module-fwdpy.slim"></span><dl class="function">
<dt id="fwdpy.slim.readslim">
<code class="descclassname">fwdpy.slim.</code><code class="descname">readslim</code><span class="sig-paren">(</span><em>filename</em><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.slim.readslim" title="Permalink to this definition">¶</a></dt>
<dd><p>Process a &#8220;SLiM&#8221; input file.</p>
<p>This function parses the input files for Phillip Messer&#8217;s &#8220;SLiM&#8221; software (<a class="reference external" href="http://messerlab.org/software/">http://messerlab.org/software/</a>).</p>
<p>Thus function only parses the following blocks in a file:
1. MUTATION RATE
2. MUTATION TYPES
3. RECOMBINATION RATE
4. GENOMIC ELEMENTS
5. CHROMOSOME ORGANIZATION</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filename</strong> &#8211; An input file for the SLiM simulation software</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">A dict containing the following data members:</td>
</tr>
</tbody>
</table>
<p>&#8216;nregions&#8217; A list of type <a class="reference internal" href="#fwdpy.fwdpy.Region" title="fwdpy.fwdpy.Region"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Region</span></code></a>.  This list represents regions where neutral muations occur.</p>
<p>&#8216;sregions&#8217; A list of types inheriting from <a class="reference internal" href="#fwdpy.fwdpy.Sregion" title="fwdpy.fwdpy.Sregion"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Sregion</span></code></a>.  This list represents regions where mutations affecting fitness occur.</p>
<p>&#8216;recregion&#8217; A list of type :class:&#8217;fwdpy.fwdpy.Region` representing recombination rate variation</p>
<p>&#8216;mu_neutral&#8217; is the total mutation rate to neutral variants (per gamete, per generation).</p>
<p>&#8216;mu_selected&#8217; is the total mutation rate to selected variants (per gamete, per generation).</p>
<p>&#8216;recrate&#8217; is the total recombination rate (per diploid, per generation).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">A dictionary containing the parsing results.</td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This function is automatically imported into fwdpy.</p>
</div>
</dd></dl>

<span class="target" id="module-fwdpy.internal.internal"></span><dl class="function">
<dt id="fwdpy.internal.internal.process_regions">
<code class="descclassname">fwdpy.internal.internal.</code><code class="descname">process_regions</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.internal.internal.process_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Process a list of objects of <code class="xref py py-class docutils literal"><span class="pre">Region</span></code></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>regions</strong> &#8211; a list of objects of <code class="xref py py-class docutils literal"><span class="pre">Region</span></code></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a pandas.DataFrame consisting of the beginning (&#8216;beg&#8217;), end (&#8216;end&#8217;) and weight (&#8216;weight&#8217;) for each elememt in l</td>
</tr>
</tbody>
</table>
<p>A user will generally not call this function.  Rather, it is used internally by things like <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.evolve_regions()</span></code></a>.</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.internal.internal.process_sregion_callbacks">
<code class="descclassname">fwdpy.internal.internal.</code><code class="descname">process_sregion_callbacks</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.internal.internal.process_sregion_callbacks" title="Permalink to this definition">¶</a></dt>
<dd><p>Process a list of objects of <code class="xref py py-class docutils literal"><span class="pre">Sregion</span></code></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>v</strong> &#8211; a <a class="reference internal" href="#fwdpy.internal.internal.shwrappervec" title="fwdpy.internal.internal.shwrappervec"><code class="xref py py-class docutils literal"><span class="pre">shwrappervec</span></code></a></li>
<li><strong>sregions</strong> &#8211; a list of objects of <code class="xref py py-class docutils literal"><span class="pre">Sregion</span></code></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">Nothing. This function populations v with necessary callbacks for the C++ code to run the desired model.</p>
</td>
</tr>
</tbody>
</table>
<p>A user will generally not call this function.  Rather, it is used internally by things like <a class="reference internal" href="#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.evolve_regions()</span></code></a>.</p>
</dd></dl>

<dl class="class">
<dt id="fwdpy.internal.internal.shwrappervec">
<em class="property">class </em><code class="descclassname">fwdpy.internal.internal.</code><code class="descname">shwrappervec</code><a class="headerlink" href="#fwdpy.internal.internal.shwrappervec" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Wrapper for a vector of callback objects from fwdpp&#8217;s extension library.</p>
<p>Users will not interact with this type directly.  Rather, it is used
by other module functions to process user inputs.</p>
</dd></dl>

<span class="target" id="module-fwdpy.fwdpyio.fwdpyio"></span><dl class="function">
<dt id="fwdpy.fwdpyio.fwdpyio.deserialize_metapops">
<code class="descclassname">fwdpy.fwdpyio.fwdpyio.</code><code class="descname">deserialize_metapops</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.deserialize_metapops" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert binary representation of populations back to a <a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>strings</strong> &#8211; A list of populations in binary format.  This should be the value returned by <a class="reference internal" href="#fwdpy.fwdpyio.fwdpyio.serialize" title="fwdpy.fwdpyio.fwdpyio.serialize"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpyio.fwdpyio.serialize()</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><a class="reference internal" href="#fwdpy.fwdpy.MetaPopVec" title="fwdpy.fwdpy.MetaPopVec"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.MetaPopVec()</span></code></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">len(strings) determines the length of the return value, and therefore the number of threads to use if the population is evolved further.</p>
</div>
<p>Example:</p>
<p>TODO</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpyio.fwdpyio.deserialize_mlocus">
<code class="descclassname">fwdpy.fwdpyio.fwdpyio.</code><code class="descname">deserialize_mlocus</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.deserialize_mlocus" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert binary representation of populations back to a <a class="reference internal" href="#fwdpy.fwdpy.MlocusPopVec" title="fwdpy.fwdpy.MlocusPopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.MlocusPopVec</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>strings</strong> &#8211; A list of populations in binary format.  This should be the value returned by <a class="reference internal" href="#fwdpy.fwdpyio.fwdpyio.serialize" title="fwdpy.fwdpyio.fwdpyio.serialize"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpyio.fwdpyio.serialize()</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><a class="reference internal" href="#fwdpy.fwdpy.MlocusPopVec" title="fwdpy.fwdpy.MlocusPopVec"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.MlocusPopVec()</span></code></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">len(strings) determines the length of the return value, and therefore the number of threads to use if the population is evolved further.</p>
</div>
<p>Example:</p>
<p>TODO</p>
</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpyio.fwdpyio.deserialize_singlepops">
<code class="descclassname">fwdpy.fwdpyio.fwdpyio.</code><code class="descname">deserialize_singlepops</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.deserialize_singlepops" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert binary representation back to a <a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopVec</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>strings</strong> &#8211; A list of populations in binary format.  This should be the value returned by <a class="reference internal" href="#fwdpy.fwdpyio.fwdpyio.serialize" title="fwdpy.fwdpyio.fwdpyio.serialize"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpyio.fwdpyio.serialize()</span></code></a></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><a class="reference internal" href="#fwdpy.fwdpy.PopVec" title="fwdpy.fwdpy.PopVec"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.PopVec()</span></code></a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">len(strings) determines the length of the return value, and therefore the number of threads to use if the population is evolved further.</p>
</div>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy.fwdpyio</span> <span class="k">as</span> <span class="nn">fpio</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">strings</span> <span class="o">=</span> <span class="p">[</span><span class="n">fpio</span><span class="o">.</span><span class="n">serialize</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="n">strings</span><span class="p">)</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops2</span> <span class="o">=</span> <span class="n">fpio</span><span class="o">.</span><span class="n">deserialize_singlepops</span><span class="p">(</span><span class="n">strings</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="fwdpy.fwdpyio.fwdpyio.gzSerializer">
<em class="property">class </em><code class="descclassname">fwdpy.fwdpyio.fwdpyio.</code><code class="descname">gzSerializer</code><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.gzSerializer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a></p>
<p>This class is a <a class="reference internal" href="#fwdpy.fwdpy.TemporalSampler" title="fwdpy.fwdpy.TemporalSampler"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.TemporalSampler</span></code></a>, allowing the state of the 
population to be written to a gzipped file at regular time points.</p>
<p>..note:: This is a good way to fill up a hard drive.  Use with caution.</p>
<dl class="method">
<dt id="fwdpy.fwdpyio.fwdpyio.gzSerializer.get">
<code class="descname">get</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.gzSerializer.get" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list for each replicate.  For each replicate, the list consists of the generation and the location of
that generation in the output file.  The latter value may be used as an offset for later reading the population
back in.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fwdpy.fwdpyio.fwdpyio.serialize">
<code class="descclassname">fwdpy.fwdpyio.fwdpyio.</code><code class="descname">serialize</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.fwdpyio.fwdpyio.serialize" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a binary representation of an evolved population</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>pop</strong> &#8211; A list of <a class="reference internal" href="#fwdpy.fwdpy.PopType" title="fwdpy.fwdpy.PopType"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.PopType</span></code></a></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">fwdpy.fwdpyio</span> <span class="k">as</span> <span class="nn">fpio</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.1</span><span class="p">),</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">ExpS</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.001</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rregions</span> <span class="o">=</span> <span class="p">[</span><span class="n">fwdpy</span><span class="o">.</span><span class="n">Region</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rng</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">GSLrng</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1000</span><span class="p">],</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">popsizes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">tile</span><span class="p">(</span><span class="n">popsizes</span><span class="p">,</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pops</span> <span class="o">=</span> <span class="n">fwdpy</span><span class="o">.</span><span class="n">evolve_regions</span><span class="p">(</span><span class="n">rng</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="n">popsizes</span><span class="p">[</span><span class="mi">0</span><span class="p">:],</span><span class="mf">0.001</span><span class="p">,</span><span class="mf">0.0001</span><span class="p">,</span><span class="mf">0.001</span><span class="p">,</span><span class="n">nregions</span><span class="p">,</span><span class="n">sregions</span><span class="p">,</span><span class="n">rregions</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">strings</span> <span class="o">=</span> <span class="p">[</span><span class="n">fpio</span><span class="o">.</span><span class="n">serialize</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">pops</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<span class="target" id="module-fwdpy.qtrait.qtrait"></span><dl class="class">
<dt id="fwdpy.qtrait.qtrait.SpopGBRTrait">
<em class="property">class </em><code class="descclassname">fwdpy.qtrait.qtrait.</code><code class="descname">SpopGBRTrait</code><a class="headerlink" href="#fwdpy.qtrait.qtrait.SpopGBRTrait" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.SpopFitness" title="fwdpy.fitness.SpopFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.SpopFitness</span></code></a></p>
<p>The &#8220;gene-based recessive&#8221; (GBR) model of Thornton et al (2013) PLoS Genetics.</p>
<p>The genetic value is the geometric mean of haplotype effect sizes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Be really careful with this one!  Fitnesses are undefined if the sum of effect sizes on a haplotype is <span class="math">\(&lt; 0:\)</span>.  The intended use case is to calculate a trait value under models with effect sizes <span class="math">\(&gt;0\)</span>.</p>
</div>
</dd></dl>

<dl class="function">
<dt id="fwdpy.qtrait.qtrait.evolve_regions_qtrait">
<code class="descclassname">fwdpy.qtrait.qtrait.</code><code class="descname">evolve_regions_qtrait</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.qtrait.qtrait.evolve_regions_qtrait" title="Permalink to this definition">¶</a></dt>
<dd><p>Evolve a quantitative trait with variable mutation, fitness effects, and recombination rates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>rng</strong> &#8211; a <code class="xref py py-class docutils literal"><span class="pre">GSLrng</span></code></li>
<li><strong>npops</strong> &#8211; The number of populations to simulate.  This is equal to the number of threads that will be used!</li>
<li><strong>N</strong> &#8211; The diploid population size to simulate</li>
<li><strong>nlist</strong> &#8211; An array view of a NumPy array.  This represents the population sizes over time.  The length of this view is the length of the simulation in generations. The view must be of an array of 32 bit, unsigned integers (see example).</li>
<li><strong>mu_neutral</strong> &#8211; The mutation rate to variants not affecting fitness (&#8220;neutral&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>mu_selected</strong> &#8211; The mutation rate to variants affecting fitness (&#8220;selected&#8221; mutations).  The unit is per gamete, per generation.</li>
<li><strong>recrate</strong> &#8211; The recombination rate in the regions (per diploid, per generation)</li>
<li><strong>nregions</strong> &#8211; A list specifying where neutral mutations occur</li>
<li><strong>sregions</strong> &#8211; A list specifying where selected mutations occur</li>
<li><strong>recregions</strong> &#8211; A list specifying how the genetic map varies along the region</li>
<li><strong>sigmaE</strong> &#8211; The standard deviation in random variation to add to trait value</li>
<li><strong>optimum</strong> &#8211; The optimum trait value. <strong>Default = 0.0</strong></li>
<li><strong>f</strong> &#8211; The selfing probabilty. <strong>Default = 0.0</strong></li>
<li><strong>VS</strong> &#8211; The total variance in selection intensity. <strong>Default = 1.0</strong></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last">RuntimeError if parameters do not pass checks</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.qtrait.qtrait.ew2010_effects">
<code class="descclassname">fwdpy.qtrait.qtrait.</code><code class="descname">ew2010_effects</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.qtrait.qtrait.ew2010_effects" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pop</strong> &#8211; A <a class="reference internal" href="#fwdpy.fwdpy.Spop" title="fwdpy.fwdpy.Spop"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fwdpy.Spop</span></code></a> simulated using <code class="xref py py-func docutils literal"><span class="pre">fwdpy.qtrait.qtrait.evolve_qtrait()</span></code> and/or <code class="xref py py-func docutils literal"><span class="pre">fwdp.qtrait.qtrait.evolve_qtrait_more()</span></code></li>
<li><strong>tau</strong> &#8211; The coupling of trait value to fitness effect of mutation</li>
<li><strong>sigma</strong> &#8211; The standard deviation for Gaussian noise applied to trait value.  Generates the <span class="math">\(\epsilon\)</span> term in E-W&#8217;s paper</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="fwdpy.qtrait.qtrait.ew2010_traits">
<code class="descclassname">fwdpy.qtrait.qtrait.</code><code class="descname">ew2010_traits</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fwdpy.qtrait.qtrait.ew2010_traits" title="Permalink to this definition">¶</a></dt>
<dd><p>Implement model of Eyre-Walker 2010</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The citation is www.pnas.org/cgi/doi/10.1073/pnas.0906182107.
We implement the simple additive case here.</p>
</div>
</dd></dl>

<span class="target" id="module-fwdpy.qtrait_mloc.qtrait_mloc"></span><dl class="class">
<dt id="fwdpy.qtrait_mloc.qtrait_mloc.MlocusGBRTrait">
<em class="property">class </em><code class="descclassname">fwdpy.qtrait_mloc.qtrait_mloc.</code><code class="descname">MlocusGBRTrait</code><a class="headerlink" href="#fwdpy.qtrait_mloc.qtrait_mloc.MlocusGBRTrait" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fwdpy.fitness.MlocusFitness" title="fwdpy.fitness.MlocusFitness"><code class="xref py py-class docutils literal"><span class="pre">fwdpy.fitness.MlocusFitness</span></code></a></p>
<p>The &#8220;gene-based recessive&#8221; (GBR) trait model of Thornton et al (2013) PLoS Genetics.</p>
<p>The GBR model is applied to each region, and then the final value is the sum over regions.</p>
<p>The fitness (or genetic value as it was used in that paper) is the geometric mean 
of haplotype effect sizes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Be really careful with this one!  Fitnesses are undefined if the sum of effect sizes on a haplotype is <span class="math">\(&lt; 0:\)</span>.  The intended use case is to calculate a trait value under models with effect sizes <span class="math">\(&gt;0\)</span>.</p>
</div>
</dd></dl>

</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
<li><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></li>
<li><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Welcome to fwdpy&#8217;s documentation!</a><ul>
<li><a class="reference internal" href="#theoretical-concepts">Theoretical concepts</a></li>
<li><a class="reference internal" href="#important-concepts-for-setting-up-simulations">Important concepts for setting up simulations</a></li>
<li><a class="reference internal" href="#examples">Examples</a></li>
<li><a class="reference internal" href="#special-topics">Special topics</a></li>
<li><a class="reference internal" href="#plugins">Plugins</a></li>
<li><a class="reference internal" href="#notes-for-people-wanting-to-work-with-the-code">Notes for people wanting to work with the code</a></li>
<li><a class="reference internal" href="#module-fwdpy.fwdpy">Functions and classes</a></li>
</ul>
</li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>

  <h4>Next topic</h4>
  <p class="topless"><a href="README.html"
                        title="next chapter">fwdpy: forward simulation in Python using fwdpp</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="README.html" title="fwdpy: forward simulation in Python using fwdpp"
             >next</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">fwdpy 0.0.4-rc3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Kevin Thornton.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.1.
    </div>
  </body>
</html>