<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Processing SLiM input files &#8212; fwdpy 0.0.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="How many threads to use?" href="threads.html" />
    <link rel="prev" title="Calculating summary statistics" href="summstats.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="processing-slim-input-files">
<h1>Processing SLiM input files<a class="headerlink" href="#processing-slim-input-files" title="Permalink to this headline">¶</a></h1>
<p><strong>fwdpy</strong> has the ability to read in the same input files as Phillip Messer&#8217;s <a class="reference external" href="http://messerlab.org/software/">SLiM</a> simulation software, which is accomplished by the function <code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.readslim()</span></code>.</p>
<p>This function does <strong>not</strong> read in all of the information in an input file.  Instead it reads in the following sections:</p>
<ul class="simple">
<li>#MUTATION RATE</li>
<li>#MUTATION TYPES</li>
<li>#GENOMIC ELEMENT TYPES</li>
<li>#CHROMOSOME ORGANIZATION</li>
<li>#RECOMBINATION RATE</li>
</ul>
<p>These blocks are parsed and converted into data types that may serve as the input parameters for functions like <a class="reference internal" href="../index.html#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.evolve_regions()</span></code></a>.</p>
<div class="section" id="differences-between-slim-and-fwdpy">
<h2>Differences between SLiM and fwdpy<a class="headerlink" href="#differences-between-slim-and-fwdpy" title="Permalink to this headline">¶</a></h2>
<p>There are several important differences between <a class="reference external" href="http://messerlab.org/software/">SLiM</a> and this package.</p>
<div class="section" id="scaling-of-fitnesses">
<h3>Scaling of fitnesses<a class="headerlink" href="#scaling-of-fitnesses" title="Permalink to this headline">¶</a></h3>
<p>For diploid genotypes <img class="math" src="../_images/math/e7425f7ac8fdd6a13108abcbce21608e0c369e3a.png" alt="AA"/>, <img class="math" src="../_images/math/1ac086ec88ca0ff7711ce23d87954bdd3e105221.png" alt="Aa"/>, and <img class="math" src="../_images/math/5a71fc441d2ecf5fd1b0c59d7181650bb5f55129.png" alt="aa"/>, <a class="reference external" href="http://messerlab.org/software/">SLiM</a> calculates fitness as <img class="math" src="../_images/math/d839e144267ecbb8a87acbc8a7dfda7824a1693e.png" alt="1"/>, <img class="math" src="../_images/math/d176a6a5c40b0ad57a1f1755451ad764229857e1.png" alt="1+sh"/>, and <img class="math" src="../_images/math/1f3dfba5a043f25cdcef2a8193cf7dc59e7cdcc3.png" alt="1+s"/>, respectively, where <img class="math" src="../_images/math/63751cb2e98ba393b0f22e45ca127c3cebb61487.png" alt="s"/> is the selection coefficient and <img class="math" src="../_images/math/293fb39e1b93282c804a86186e721b32f829f1b2.png" alt="h"/> the dominance associated with the <img class="math" src="../_images/math/bd9d9c7512c898e0a254a0882184197bc1bd9c9f.png" alt="a"/> allele.  In <strong>fwdpy</strong>, the fitnesses of the three genotypes are <img class="math" src="../_images/math/d839e144267ecbb8a87acbc8a7dfda7824a1693e.png" alt="1"/>, <img class="math" src="../_images/math/d176a6a5c40b0ad57a1f1755451ad764229857e1.png" alt="1+sh"/>, and <img class="math" src="../_images/math/3c944f09cd56b4b4ced1e7449af275d11ca8d3c3.png" alt="1+2s"/>, respectively.   To account for this difference, <code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.readslim()</span></code> halves the selection coefficient (or the mean <img class="math" src="../_images/math/63751cb2e98ba393b0f22e45ca127c3cebb61487.png" alt="s"/> for distributions on selection coefficients) and doubles the dominance found in the #MUTATION TYPES block of a <a class="reference external" href="http://messerlab.org/software/">SLiM</a> input file.</p>
<p><strong>IMPORTANT:</strong> <a class="reference external" href="http://messerlab.org/software/">SLiM</a> input files should be generated according to that software&#8217;s manual!!!  In other words, co-dominance should be coded as <img class="math" src="../_images/math/8e2d2a423de595e5db6c11816d70fde7f3eac372.png" alt="h=0.5"/> and you should allow <code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.readslim()</span></code> to convert it to one automatically!  Further, the fitnesses of the three genotypes should be treated as  <img class="math" src="../_images/math/d839e144267ecbb8a87acbc8a7dfda7824a1693e.png" alt="1"/>, <img class="math" src="../_images/math/d176a6a5c40b0ad57a1f1755451ad764229857e1.png" alt="1+sh"/>, and <img class="math" src="../_images/math/1f3dfba5a043f25cdcef2a8193cf7dc59e7cdcc3.png" alt="1+s"/>, respectively. The intent here is to allow pre-existing <a class="reference external" href="http://messerlab.org/software/">SLiM</a> input files to be used in <strong>fwdpy</strong> with no additional modification.</p>
</div>
<div class="section" id="discrete-vs-continuous-positions">
<h3>Discrete vs continuous positions<a class="headerlink" href="#discrete-vs-continuous-positions" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="http://messerlab.org/software/">SLiM</a>&#8216;s mutation model involves <em>discrete</em> positions along the genome.  Therefore, the output from that software contains mutations with postitions whose values are integers.  Likewise, recombination events occur between integer-valued positions in <a class="reference external" href="http://messerlab.org/software/">SLiM</a>.</p>
<p>Internally, this package uses <a class="reference external" href="http://github.com/molpopgen/fwdpp">fwdpp</a> to model mutation and recombination positions as <em>continuous</em> random variables along genomic intervals in a fashion similar to the coalescent simulation <a class="reference external" href="http://home.uchicago.edu/~rhudson1">ms</a>.  (Techincally, <a class="reference external" href="http://home.uchicago.edu/~rhudson1">ms</a> treats mutation as continuous and recombination breakpoints as discrete, but let&#8217;s ignore that point for now.)</p>
<p>An additional difference is that <a class="reference external" href="http://messerlab.org/software/">SLiM</a> is paramterized in terms of &#8220;per base pair&#8221; mutation and recombination rates and this package uses &#8220;per region&#8221; rates.</p>
<p><a class="reference internal" href="../index.html#fwdpy.fwdpy.evolve_regions" title="fwdpy.fwdpy.evolve_regions"><code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.evolve_regions()</span></code></a> adjusts the <a class="reference external" href="http://messerlab.org/software/">SLiM</a> input file to account for these differences.  Let&#8217;s consider the following setup for recombination rates:</p>
<blockquote>
<div><div class="line-block">
<div class="line">#RECOMBINATION RATE</div>
<div class="line">10000 1e-7</div>
<div class="line">20000 2e-7</div>
<div class="line">30000 3e-7</div>
</div>
</div></blockquote>
<p>The block above means that genomic positions 1 through 10,000 have a recombination rate of <img class="math" src="../_images/math/aa1f42b09a4460a6c0ac82b3909f879b96de5a29.png" alt="10^{-7}"/> <em>per base pair</em> (bp).  Positions 10,001 through 20,000 have a recombination rate of <img class="math" src="../_images/math/b0a66d69da43d13ca2cf3388c03a30dd0d749597.png" alt="2 \times 10^{-7}/\mathrm{bp}"/>, and positions 20,001 through 30,000 have a recombination rate of <img class="math" src="../_images/math/d6be9ce9d2d097d7248063afd1ece102a53bdcd6.png" alt="3 \times 10^{-7}/\mathrm{bp}"/>.  The total/cumulative recombination rate is <img class="math" src="../_images/math/ac9e0940e2e97116ea33f9a33699adc7ff7dd5e5.png" alt="10^4\times(10^{-7} + 2\times 10^{-7} + 3 \times 10^{-7}) = 0.006"/>
The function <code class="xref py py-func docutils literal"><span class="pre">fwdpy.fwdpy.readslim()</span></code> will  convert the above block to three half-open, continuous intervals:</p>
<blockquote>
<div><div class="line-block">
<div class="line"><img class="math" src="../_images/math/e46fb243bd11cbd80d4a0c53fbbd264baaaeba65.png" alt="[0,10000)"/>, with <em>total</em> recombination rate <img class="math" src="../_images/math/79a40dffa033e35efcd3442006ebd8c43a629ff1.png" alt="10^{-7}\times(10^4-0) = 10^{-3}"/></div>
<div class="line"><img class="math" src="../_images/math/9d88c5aa5d289a6dcb3dbadf23e27a2f603ef53f.png" alt="[10000,20000)"/> with <em>total</em> recombination rate <img class="math" src="../_images/math/fa9434bae21e483b370f70c2b64057bfebf9308e.png" alt="2 \times 10^{-7}\times(2\times 10^4-10^4) = 2\times10^{-3}"/></div>
<div class="line"><img class="math" src="../_images/math/5dd7bb2d6fb8054ccb073b81eaf046450b1072da.png" alt="[20000,30000)"/> with <em>total</em> recombination rate <img class="math" src="../_images/math/06f8e23db2fc19fb922e41d86c1a4419e7c20761.png" alt="3 \times 10^{-7}\times(3\times 10^4-2\times 10^4) = 3\times10^{-3}"/></div>
</div>
</div></blockquote>
<p>You can see that the total recombination rate is the same (<img class="math" src="../_images/math/ddfcfe7c91b7eeefcccccd00843940c953190b2a.png" alt="0.006"/>).  The difference from <a class="reference external" href="http://messerlab.org/software/">SLiM</a> is that crossover positions will be drawn from continuous uniform distributions.</p>
<p>The intervals where mutations occur are treated similarly.  Further, for a given genomic element, the mutation rate to a particular mutation type is the mutation rate specified in the #MUTATION RATE block times the weight assigned to that mutation type divided by the sum of all weights for all mutation types occuring in that genomic element.</p>
</div>
</div>
<div class="section" id="a-word-of-caution">
<h2>A word of caution<a class="headerlink" href="#a-word-of-caution" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="http://messerlab.org/software/">SLiM</a> allows different mutations to have the same position label.  This software does not.  This difference may be relevant when comparing simulation output to theoretical predictions, which typically come from the infinitely-many sites assumption that this package models.  For a region of length <img class="math" src="../_images/math/ae2b750f71e1fc0daaa3de9a85d42794d7cd1326.png" alt="L"/>, as <img class="math" src="../_images/math/d8540ed3f8c98077eeb7ecb41df34e42a979f13b.png" alt="\mu_{bp}"/> (the mutation rate per base pair) the output of <a class="reference external" href="http://messerlab.org/software/">SLiM</a> and <strong>fwdpy</strong> should be comparable for large <img class="math" src="../_images/math/ae2b750f71e1fc0daaa3de9a85d42794d7cd1326.png" alt="L"/> and <img class="math" src="../_images/math/c3599fd9de58c300d80f82a4538b9525f2c164f1.png" alt="\mu_{bp} \to 0"/>.  When mutation rates are large relative to <img class="math" src="../_images/math/ae2b750f71e1fc0daaa3de9a85d42794d7cd1326.png" alt="L"/>, <a class="reference external" href="http://messerlab.org/software/">SLiM</a>&#8216;s mutation scheme becomes a quasi-finite-sites model.</p>
</div>
<div class="section" id="limitations">
<h2>Limitations<a class="headerlink" href="#limitations" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Gene conversion is not currently supported</li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Processing SLiM input files</a><ul>
<li><a class="reference internal" href="#differences-between-slim-and-fwdpy">Differences between SLiM and fwdpy</a><ul>
<li><a class="reference internal" href="#scaling-of-fitnesses">Scaling of fitnesses</a></li>
<li><a class="reference internal" href="#discrete-vs-continuous-positions">Discrete vs continuous positions</a></li>
</ul>
</li>
<li><a class="reference internal" href="#a-word-of-caution">A word of caution</a></li>
<li><a class="reference internal" href="#limitations">Limitations</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="summstats.html" title="previous chapter">Calculating summary statistics</a></li>
      <li>Next: <a href="threads.html" title="next chapter">How many threads to use?</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/pages/slim.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, KRT.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.9</a>
      
      |
      <a href="../_sources/pages/slim.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>